### PR TITLE
#2438: Auto-focus first enabled field in modal form

### LINKED ISSUE #2438: Auto-focus first enabled field in modal form
Context
- As a user, I want to trigger a form with the quickbar, and not have to use my mouse to fill it out
- This requires the field being initially focused, or being able to hit tab to focus the first editable field

Related Discussion:
- https://pixiebrix.slack.com/archives/C02CN01JXAA/p1642648962063600

### PR DESCRIPTION
**Pull request by fregante**
- Closes #2438

## Before

https://user-images.githubusercontent.com/1402241/151548210-be962f1e-d20e-4de0-9c22-6f41116cc62a.mov

## After

https://user-images.githubusercontent.com/1402241/151548212-415e2276-c356-4545-b337-a789f3532008.mov

### CODE REVIEW

File: src/blocks/transformers/ephemeralForm/EphemeralForm.tsx [hunk 1]
```diff
@@ -33,19 +33,24 @@ const uiWidgets = {
...
   </div>
-);
+));
+ModalLayout.displayName = "ModalLayout";
 
-const PanelLayout: React.FC = ({ children }) => <div>{children}</div>;
+const PanelLayout = forwardRef<HTMLDivElement>((props, ref) => (
+  <div ref={ref}>{props.children}</div>
+));
+PanelLayout.displayName = "PanelLayout";
```
Comment (fregante) [inline src/blocks/transformers/ephemeralForm/EphemeralForm.tsx:47]: https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/display-name.md

File: src/blocks/transformers/ephemeralForm/EphemeralForm.tsx [hunk 2]
```diff
@@ -33,19 +33,24 @@ const uiWidgets = {
   imageCrop: ImageCropWidget,
 };
 
-const ModalLayout: React.FC = ({ children }) => (
+const ModalLayout = forwardRef<HTMLDivElement>((props, ref) => (
```
Comment (fregante) [inline src/blocks/transformers/ephemeralForm/EphemeralForm.tsx:36]: This appears to be the right way to pass references around, but it's causing a type error I don't understand. Help?

https://reactjs.org/docs/forwarding-refs.html

```

src/blocks/transformers/ephemeralForm/EphemeralForm.tsx:78:8 - error TS2559: Type '{ children: Element; }' has no properties in common with type 'IntrinsicAttributes & RefAttributes<HTMLDivElement>'.

78       <FormContainer>
          ~~~~~~~~~~~~~

src/blocks/transformers/ephemeralForm/EphemeralForm.tsx:86:8 - error TS2559: Type '{ children: Element; }' has no properties in common with type 'IntrinsicAttributes & RefAttributes<HTMLDivElement>'.

86       <FormContainer>
          ~~~~~~~~~~~~~

src/blocks/transformers/ephemeralForm/EphemeralForm.tsx:93:6 - error TS2322: Type '{ children: Element; ref: RefObject<HTMLDivElement>; }' is not assignable to type 'IntrinsicAttributes & RefAttributes<HTMLDivElement>'.
  Property 'children' does not exist on type 'IntrinsicAttributes & RefAttributes<HTMLDivElement>'.

93     <FormContainer ref={form}>
        ~~~~~~~~~~~~~

[7:42:52 PM] Found 3 errors. Watching for file changes.
```

File: src/blocks/transformers/ephemeralForm/EphemeralForm.tsx [hunk 3]
```diff
@@ -62,6 +67,12 @@ const EphemeralForm: React.FC = () => {
     [nonce]
   );
 
+  useEffect(() => {
+    form.current
+      ?.querySelector<HTMLInputElement | HTMLSelectElement>("input, select")
+      ?.focus();
+  });
```
Comment (fregante) [inline src/blocks/transformers/ephemeralForm/EphemeralForm.tsx]: Solution also used in https://github.com/pixiebrix/pixiebrix-extension/blob/e51971a2f17e7dc19a54bf687710106536bf85f9/src/devTools/editor/tabs/editTab/editorNode/EditorNode.tsx#L89-L93

It doesn't look like `JsonSchemaForm` itself has any way to access the generated elements https://github.com/rjsf-team/react-jsonschema-form/blob/3c1d4659580a618a0395b4ea916710b1e3c258a0/packages/core/index.d.ts#L11-L53

**Comment by twschiller** [comment]
Adding @BALEHOK as reviewer since he's expert on all things form builder

**Comment by fregante** [comment]
I'm not familiar with React’s lifecycle so I can't answer those.

If the underlying component already supports `autofocus` as you say then we should probably use that since they probably dealt with this kind of issues. I'm not sure how to go about dynamically adding the autofocus to the first item of the definition though.

**Comment by fregante** [comment]
I'll close this PR for now as the solution is likely far from what I pushed

**Pull request closed without merging**